Result File Dataclasses

mawelborn opened the pull request:
This PR adds a new set of dataclasses for working with result files as native Python types rather than nested `dict`s and `list`s.

- Python 3.8+ compatible, fully type checked, and high unit test coverage
- Supports result file versions 1 and 3 with a consistent API
- Supports loading results from a `dict`, JSON `str`, or `Path`
- Provides an API for sorting, filtering, grouping, and modifying predictions including:
    - `predictions.where()`
    - `predictions.groupby()`
    - `predictions.orderby()`
    - `predictions.apply()`
- Provides functionality to accept/reject predictions and produce a `changes` dict suitable for `SubmitReview`.

See [the example script](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/result-file-dataclasses/examples/result_file_dataclasses.py) for sample use cases and attribute references.

The dataclasses have been added as a new `indico_toolkit.results` module rather than replacing `indico_toolkit.types` as originally planned. The existing `types` module is used in many places, and updating its dependents would make this already-large PR much bigger. As a standalone PR this will be easier to review, and `types` can be replaced piecemeal as needed in future PRs.

Scott771 commented:
Looks like there is an issue with one of the datasets utilized in a fixture (possibly got deleted in some try migration recently?) as well as a failed fileprocessing test

Scott771 commented on examples/result_file_dataclasses.py:
```diff
@@ -0,0 +1,202 @@
+from operator import attrgetter
+from pathlib import Path
+
+import pendulum
```
if we're not adding polars / pendulum as dependencies to the repo, I'd consider removing those from the examples file (or commenting that these are optional)

mawelborn replied:
I've removed these from the examples, opting to use a simpler string-based normalization rule and native CSV writer instead.

Scott771 commented on examples/result_file_dataclasses.py:
```diff
@@ -0,0 +1,202 @@
+"""
+
+
+def get_predictions(result: results.Result) -> Iterator[dict[str, object]]:
```
I'd add in the assumption here-- i.e. `assumes you have one classification model and one set of extraction results`

mawelborn replied:
The updated `results_to_csv.py` script now makes no assumptions on the number of classification or extraction models and instead dumps all of them.

Scott771 commented on examples/result_file_dataclasses.py:
```diff
@@ -0,0 +1,202 @@
+        }
+
+
+pl.DataFrame(get_predictions(result)).write_csv("predictions.csv")
```
curious what your thoughts are on putting this functionality into the class itself?

mawelborn replied:
I'm hesitant to add more functionality to the dataclasses, as they're already quite large. My guiding principle was to create the kind of result API that I'd like to use to write higher level code like this.

I've been thinking that it might be better to group that kind of functionality in other modules. For example, a downselection function that uses `groupby()`, `orderby()` and `reject()` underneath would makes sense to have in the auto review module. And that keeps these classes from ending up with all manner of auto review, post processing, and prediction dump code in them.

Of course I'm still very open to adding new functions to the classes if they'd be generally useful to all those cases--things we find ourselves writing repeatedly and would be nice to have generalized as a part of the result API itself.

Scott771 replied:
makes sense!

Scott771 commented on examples/result_file_dataclasses.py:
```diff
@@ -0,0 +1,202 @@
+extractions = pre_review.extractions
+
+# Downselect based on highest confidence.
+for model, extractions in extractions.groupby(attrgetter("model")).items():
```
for the examples, I think we're getting to customized in the behavior here to be generally useful (unless someone is reproducing this exact code)

Scott771 replied:
or perhaps consider incorporating `attrgetter` into the class itself

mawelborn replied:
I've split the example script out into three examples covering results to CSV, a minimal auto review, and loading and traversing results. Hopefully that makes them more focused and useful as real executable examples.

Highest confidence downselection was just one of the common use cases for auto review that also showed off quite a bit of the results API. Definitely open to ideas for the auto review example that are less customized!

Scott771 commented on examples/result_file_dataclasses.py:
```diff
@@ -0,0 +1,202 @@
+for label, threshold in thresholds.items():
+    extractions.where(label=label, min_confidence=threshold).accept()
+
+extractions.where(max_confidence=0.75).reject()
```
I'd comment above that you're rejecting for all predictions here

mawelborn commented:
> Looks like there is an issue with one of the datasets utilized in a fixture (possibly got deleted in some try migration recently?) as well as a failed fileprocessing test

Looks like the tests had been updated on `main`. Merging that in seems to have fixed it! Also refactored the example script and split it into three in an attempt to make each more individually useful.

mawelborn commented:
Couple quick updates:

- Added `document.full_text_url` alongside the existing `document.etl_output_url`.
- Fixed the calculation for `result.rejected` to account for admin reviews.
- Made `extraction.text` work with 6.6+ Typed Entities changes.

For the latter, I designed the API to use the `text` attribute consistently across versions. It does what you want it to do 99% of the time while making Typed Entities data available if needed.

- Getting `extraction.text` returns the OCR text from the predicted span.
- Setting `extraction.text` will update in the output of `to_changes()`:
    - `prediction["text"]` for 5.X-6.5 compatibility
    - `prediction["normalized"]["text"]` for consistency
    - `prediction["normalized"]["formatted"]` for 6.6+ compatibility
- If needed, Typed Entities data is available alongside all other not-explicitly-parsed data in the `extras` attribute:
    - `extraction.extras["normalized"]["formatted"]`

mawelborn commented:
I've done a significant rewrite of this PR in light of recent changes to the platform. It now aims to provide a consistent API matching closely to our v4 result file proposal. It uses IPA 6.8+ version 3 as the new baseline for result files with support for IPA 6.8+ version 1 (sans some attributes not present in v1) due to their ubiquity. It does not attempt to support version 2 result files or the various inconsistencies of IPA 6.7 and prior due to their limited use.

Overall the code and the API are simpler and cleaner. I've dogfooded it in a couple projects and think it's ready for broader use and feedback.
